### PR TITLE
For #27472 - Added telemetry data points for the Google Play Store review prompts

### LINKED ISSUE #27472: Capture telemetry for Google Play Store Review prompts
[comment]: # (Please do your best to search for duplicate issues before filing a new issue so we can keep our issue board clean)
[comment]: # (Every issue should have exactly one feature request described in it. Please do not file feedback list tickets as it is difficult to parse them and address their individual points)
[comment]: # (Feature Requests are better when they’re open-ended instead of demanding a specific solution e.g: “I want an easier way to do X” instead of “add Y”)
[comment]: # (Read https://github.com/mozilla-mobile/fenix#i-want-to-file-an-issue for more information)
We currently occasionally ask our users to rate our app in the Play Store via a prompt. There have been some reports from our users that we repeatedly ask them to review the app and that the prompt appears at odd hours. 

### What is the user problem or growth opportunity you want to see solved?
We want to understand and capture actual data on this app feature and behavior to better understand what is happening in the wild. 
- We need data on when and how often we ask our users to rate the app
- How many people rate the app or ignore the prompt

### How do you know that this problem exists today? Why is this important?
User reports thru SUMO. 

### Who will benefit from it?
The data will help us validate user reports and make adjustments in our implementation.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21834)

### PR DESCRIPTION
**Pull request by t-p-white**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [ ] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #27472

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/components/ReviewPromptController.kt [hunk 1]
```diff
@@ -46,14 +48,24 @@ class ReviewPromptController(
...
+        )
+        ReviewPrompt.lastPromptDatetime.set(Date(reviewSettings.lastReviewPromptTimeInMillis))
+
         withContext(Main) {
             flow.addOnCompleteListener {
                 if (it.isSuccessful) {
                     manager.launchReviewFlow(activity, it.result)
+                    ReviewPrompt.promptAttempted.set(Date())
+                } else {
+                    ReviewPrompt.promptNotAttempted.set(Date())
```
Comment (t-p-white) [inline app/src/main/java/org/mozilla/fenix/components/ReviewPromptController.kt:64]: note to self - remove

File: app/metrics.yaml [hunk 2]
```diff
@@ -8439,3 +8439,31 @@ recent_synced_tabs:
...
+    type: event
+    description: |
+      Data captured for each attempt to display the review prompt.
+    extra_keys:
+      prompt_was_displayed:
+        description: |
+          Whether the prompt was displayed to the user. Value
+          reported will be: 'true', 'false' or 'error'.
+        type: string
+      now_local_datetime:
```
Comment (rocketsroger) [inline app/metrics.yaml:8453]: I suggest to use `local_datetime`.  The word "now" might be confusing.

File: app/metrics.yaml [hunk 3]
```diff
@@ -8439,3 +8439,31 @@ recent_synced_tabs:
...
+        description: |
+          The 'now' local datetime.
```
Comment (rocketsroger) [inline app/metrics.yaml:8455]: Lets remove the 'now' here as well.

File: app/src/main/java/org/mozilla/fenix/components/ReviewPromptController.kt [hunk 4]
```diff
@@ -98,3 +107,44 @@ class ReviewPromptController(
...
+        "false"
+    } else if (reviewInfoAsString.contains("isNoOp=false")) {
+        "true"
+    } else {
+        "error"
+    }
+
+    ReviewPrompt.promptAttempt.record(
+        ReviewPrompt.PromptAttemptExtra(
+            nowLocalDatetime = formattedNowLocalDatetime,
```
Comment (rocketsroger) [inline app/src/main/java/org/mozilla/fenix/components/ReviewPromptController.kt:145]: nit: can we please update the calls here to match the order in metrics.yaml?

**Review by rocketsroger — COMMENTED** [review]
Looks good.  Small nits

**Comment by t-p-white** [comment]
# Request for data collection review form
All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
1. What questions will you answer with this data?
  * The data will help us understand why some users are experiencing sporadic and repeated prompts to review the app.
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
  * We are adding telemetry data in order to identify when repeated review prompts occur - which cause a detrimental experience to the user.
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
  * There are no other alternatives. 
4. Can current instrumentation answer these questions?
  * No. 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki. 
Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.

Measurement Description | Data Collection Category | Tracking Bug #
--- | --- | ---
Prompt user to review: The user has been prompted to review the application. | Category 2 - interaction data |  https://github.com/mozilla-mobile/fenix/issues/27472

6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
  * https://dictionary.telemetry.mozilla.org/apps/fenix 
7. How long will this data be collected? Choose one of the following:
  * Up until version 121, with the option to renew at that point.
8. What populations will you measure?
  * All channels, all locales, all countries 
9. If this data collection is default on, what is the opt-out mechanism for users?
  * Default Glean SDK opt-out mechanism. 
10. Please provide a general description of how you will analyze this data.
  * Glean and Glam. 
11. Where do you intend to share the results of your analysis?
  *  Only on Glean, Glam, and with mobile teams. 
12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
   * No third-party tools.

**Comment by rocketsroger** [comment]
> # Request for data collection review form
> 
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
>     1. What questions will you answer with this data?
> 
> 
>     * The data will help us understand why some users are experiencing sporadic and repeated prompts to review the app.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
>     * We are adding telemetry data in order to identify when repeated review prompts occur - which cause a detrimental experience to the user.
> 
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
>     * There are no other alternatives.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * No.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.    
>        Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> Prompt user to review: The user has been prompted to review the application. 	Category 2 - interaction data 	#27472
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
>     * [https://dictionary.telemetry.mozilla.org/apps/fenix   ](https://dictionary.telemetry.mozilla.org/apps/fenix%E2%80%A8)
> 
> 
>     7. How long will this data be collected? Choose one of the following:
> 
> 
>     * Up until version 121, with the option to renew at that point.
> 
> 
>     8. What populations will you measure?
> 
> 
>     * All channels, all locales, all countries
> 
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
>     * Default Glean SDK opt-out mechanism.
> 
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
>     * Glean and Glam.
> 
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
>     * Only on Glean, Glam, and with mobile teams.
> 
> 
>     12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
> 
> 
>     * No third-party tools.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection set to end or be renewed by version 121

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

File: app/src/main/java/org/mozilla/fenix/components/ReviewPromptController.kt [hunk 5]
```diff
@@ -6,10 +6,14 @@ package org.mozilla.fenix.components
...
 import android.app.Activity
 import androidx.annotation.VisibleForTesting
+import com.google.android.play.core.review.ReviewInfo
 import com.google.android.play.core.review.ReviewManager
 import kotlinx.coroutines.Dispatchers.Main
 import kotlinx.coroutines.withContext
 import org.mozilla.fenix.utils.Settings
+import org.mozilla.fenix.GleanMetrics.ReviewPrompt
+import java.text.SimpleDateFormat
+import java.util.*
```
Comment (rocketsroger) [inline app/src/main/java/org/mozilla/fenix/components/ReviewPromptController.kt:16]:
```suggestion
import java.util.Date
import java.util.Locale
```

**Review by rocketsroger — APPROVED** [review]
Looks good.  Suggest that we remove the wildcard import.

**Review by rocketsroger — APPROVED** [review]
🚢

**Pull request merged**